text/0000-command-ergonomics.md

```diff
@@ -0,0 +1,656 @@
+```
+
+No significant changes are made to the `Command` construction APIs,
+but it may become necessary to call `.stderr()` explicitly
+to use the new methods (see Unresolved Questions).
+
+## Use cases
+
+We aim to serve well each of the following people:
+
+ * Alice writes a CLI utility to orchestrate processes and wants top-notch error reporting from subprocesses.
+
+ * Bob migrates ad-hoc automation scripts from bash to Rust.
+
+ * Carol writes a generic application and just wants all errors to be maximally useful by default.
+
+ * Dionysus wants to run `diff`, which exits `0` for "no difference",
+   `1` for "difference found" and
+   another value for failure.
+
+(Partly cribbed from one of
+@matklad's [comments](https://github.com/rust-lang/rust/pull/89004#issuecomment-923803209) in #89004)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## New methods on `Command`:
+
+ * `fn run(&mut self) -> Result<(), ProcessError>`:
+
+   Runs the command.
+   Equivalent to `.spawn()` followed by `.status()`,
+   but with better error handling.
+
+ * `fn read_stdout(&mut self) -> Result<Vec<u8>, ProcessError>`:
+
+   Runs the command and collects its stdout.
+   After the child indicates EOF on its stdout,
+   we will wait for it to finish and check the exit status.
+
+ * `fn read_stdout_string(&mut self) -> Result<String, ProcessError>`:
+
+   Runs the command and collects its stdout, as for `read_stdout`.
+   Decodes the stdout as UTF-8, and fails if that's not possible.
+   Does not trim any trailing line ending.
+
+ * `fn stdout_reader(&mut self) -> std::process::ChildOutputReader`
+   (where `struct ChildOutputReader` implements `io::Read`
+   and is `Send + Sync + 'static`).
+
+   Starts the command, allowing the caller to
+   read the stdout in a streaming way.
+   Neither EOF nor an error will be reported by `ChildOutputReader`
+   until the child has exited, *and* the stdout pipe reports EOF.
+   (This includes errors due to nonempty stderr,
+   if stderr was set to `piped`.)
+
+Most callers should use these methods,
+rather than `.spawn()` or `.status()`.
+These new methods should be be
+recommended by docs for other, lower-level functions
+which people were previously required to use.
+
+## Deprecations
+
+ * Apply `#[must_use]` to `Command` and `Child`.
+
+ * Apply `#[must_use]` to `ExitStatus`.
+   (May require fixing quite a few of the examples.)
+
+ * Add a warning to the docs for `Command.output()` about the lost error bugs,
+   in particular the need to check `.status` and the lack of any compiler
+   warning if one doesn't.  Suggest to the reader to consider
+   `.run()` or `.read_stdout*` instead.
+   Do not deprecate `.output()` though.
+
+## stderr handling
+
+If `stderr(Stdio::piped())`,
+these new functions all collect the child's stderr.
+Then,
+if the stderr output is nonempty, this is considered an error,
+and reported in the `ProcessError`.
```

Comment:
I'm a bit uncertain about this semantics.  It's popular in command-line applications to use `stderr` to collect running commentary about what the program is doing (for example `log::debug` under `env_logger`) in ways that definitely shouldn't be considered errors.

Reply:
I have an experience report that erroring out on non-empty stderr is wrong! One common case here is when you run something like `rustc --version`, but that thing is `rustup` managed, so, before you get to `rustc --version` output, you have printout from `rustup` on the stderr that it downloads the appropriate version.